File: stoat/list_snarl_paths.py
import bdsg # type: ignore
import argparse
from stoat import utils
import time 
import os 
from typing import List, Tuple

# class to help make paths from BDSG objects
# and deal with orientation, flipping, etc
class Path:
    def __init__(self):
        self.nodes = []
        self.orients = []

    def addNode(self, node, orient):
        # here we know the actual node id and orientation
        self.nodes.append(node)
        self.orients.append(orient)

    def addNodeHandle(self, node_h, stree):
        # we have a BDSG handle and need to extract node id and orientation
        # couldn't find a better way than parsing the
        # string representation of the node...
        node_s = stree.net_handle_as_string(node_h)
        # trivial chain
        if stree.is_trivial_chain(node_h):
            node_s = node_s.replace(' pretending to be a chain', '')
            node_s = node_s.replace(' in a simple snarl', '')

        # parse node info
        node_s = node_s.replace('node ', '')
        node_o = '>'
        if 'rev' in node_s:
            node_o = '<'
        node_s = node_s.replace('rev', '').replace('fd', '')
        # add node to path
        self.nodes.append(node_s)
        self.orients.append(node_o)

    def print(self):
        # write the string representation of the path
        # e.g. ">I>J<K>L" or ">I>J>*>M>N"
        out_path = []
        for ii in range(len(self.nodes)):
            out_path.append(self.orients[ii] + str(self.nodes[ii]))
        return (''.join(out_path))

    def flip(self):
        # some paths are traversing the nodes entirely or mostly in reverse
        # for convenience we might can to flip them
        self.nodes.reverse()
        self.orients.reverse()
        for ii in range(len(self.orients)):
            if self.nodes[ii] == '*':
                # don't flip the orientation of the * because
                # it should be ">" by construction
                continue
            if self.orients[ii] == '>':
                self.orients[ii] = '<'
            else:
                self.orients[ii] = '>'

    def size(self):
        return (len(self.nodes))

    def nreversed(self):
        # counts how many nodes are traversed in reverse
        return (sum(['<' == orient for orient in self.orients]))

def calcul_pos_type_variant(list_list_length_paths: List[List[str]]) -> Tuple[List[str], int]:
    list_type_variant = []
    padding = 0
    just_snp = True

    for path_lengths in list_list_length_paths:
        if len(path_lengths) > 3 or path_lengths[1] == "_":  # Case snarl in snarl / Indel
            list_type_variant.append("CPX")  # COMPLEX
            just_snp = False
        elif len(path_lengths) == 3:  # Case simple path len 3
            if len(path_lengths[1]) == 1:
                list_type_variant.append(path_lengths[1])  # add node str snp
            else:
                ins_seq = "INS" if len(path_lengths[1]) > 3 else path_lengths[1]
                list_type_variant.append(ins_seq)
                just_snp = False
        elif len(path_lengths) == 2:  # Deletion
            list_type_variant.append("DEL")
            just_snp = False
        elif not path_lengths:  # Case path_lengths is empty
            ValueError("path_lengths is empty")

    # add +1 in pos for just SNP present in snarl
    if just_snp:
        padding = 1

    return list_type_variant, padding

def check_threshold(proportion) :
    proportion = float(proportion)
    if proportion <= 0 :
        raise argparse.ArgumentTypeError("Proportion value must be >0.")

    return proportion

def find_snarl_id(stree, snarl) :
    # create a snarl ID as LEFT_RIGTH bondary nodes
    sstart = stree.get_bound(snarl, False, True)
    sstart = stree.get_node_from_sentinel(sstart)
    send = stree.get_bound(snarl, True, True)
    send = stree.get_node_from_sentinel(send)
    snarl_id = '{}_{}'.format(stree.node_id(send), stree.node_id(sstart))

    return snarl_id

def follow_edges(stree, finished_paths, path, paths, pg) :
    def add_to_path(next_child) :

        if stree.is_sentinel(next_child):
            # If this is the bound of the snarl then we're done
            # Because we only traverse in the netgraph, it can only be the
            # bound of the parent snarl
            finished_paths.append([])
            for net in path:
                finished_paths[-1].append(net)
            finished_paths[-1].append(next_child)
        else :
            for i in path : 
                # Case where we find a loop 
                if stree.net_handle_as_string(i) == stree.net_handle_as_string(next_child) :
                    return False
            paths.append([])
            for net in path:
                paths[-1].append(net)
            paths[-1].append(next_child)
        return True

    # from the last thing in the path
    stree.follow_net_edges(path[-1], pg, False, lambda n: add_to_path(n))
 
def save_snarls(stree, root, pg, ref_paths, ppo) :
    snarls = []
    snarls_pos = {}

    # given a node handle (dist index) return a position on a reference path
    def get_node_position(node):
        node_h = stree.get_handle(node, pg)
        ret_pos = []

        def step_callback(step_handle):
            path_handle = pg.get_path_handle_of_step(step_handle)
            path_name = pg.get_path_name(path_handle)

            if path_name in ref_paths:
                position = int(ppo.get_position_of_step(step_handle) + stree.node_length(node))
                ret_pos.append(path_name)
                ret_pos.append(position)
                return False
            return True

        pg.for_each_step_on_handle(node_h, step_callback)
        return (ret_pos)

    def save_snarl_tree_node(net):
        # check if we can find a position for this snarl on the reference path
        snarl_pos = get_net_start_position(net)
        # if we couldn't find a position, use the parent's that we should have
        # found and saved earlier
        if len(snarl_pos) == 0:
            par_net = stree.get_parent(net)
            snarl_pos = snarls_pos[stree.net_handle_as_string(par_net)]
        # save this position
        snarls_pos[stree.net_handle_as_string(net)] = snarl_pos
        # save snarl
        if stree.is_snarl(net):
            snarls.append([net, snarl_pos[0], snarl_pos[1]])
        # explore children
        if not stree.is_node(net) and not stree.is_sentinel(net):
            stree.for_each_child(net, save_snarl_tree_node)
        return (True)

    # given a handle for a node, snarl, chain, whatever, return a "start" position
    def get_net_start_position(net):
        # if node, look at its position
        if stree.is_node(net):
            return get_node_position(net)
        # otherwise check boundaries
        bnode1 = stree.get_bound(net, True, False)
        bnode1_p = get_node_position(bnode1)
        bnode2 = stree.get_bound(net, False, False)
        bnode2_p = get_node_position(bnode2)

        # if one of the bondaries is not on a reference path, return it
        if len(bnode1_p) == 0:
            return bnode1_p
        if len(bnode2_p) == 0:
            return bnode2_p
        assert bnode1_p[0] == bnode2_p[0], 'boundary nodes on different reference paths'
        # as "start" position, let's return the smallest position
        if bnode1_p[1] < bnode2_p[1]:
            return bnode1_p
        else:
            return bnode2_p
    
    stree.for_each_child(root, save_snarl_tree_node)
    return snarls

def parse_graph_tree(pg_file, dist_file) :

    # load graph and snarl tree
    pg = bdsg.bdsg.PackedGraph()
    pg.deserialize(pg_file)
    stree = bdsg.bdsg.SnarlDistanceIndex()
    stree.deserialize(dist_file)
    pp_overlay = bdsg.bdsg.PackedPositionOverlay(pg)

    # list all snarls in pangenome
    # init with the child (only one ideally) of the root
    root = stree.get_root()
    return stree, pg, root, pp_overlay

def parse_pg(pg_file) :

    # load graph and snarl tree
    pg = bdsg.bdsg.PackedGraph()
    pg.deserialize(pg_file)
    return pg

def fill_pretty_paths(stree, pg, finished_paths) :
    pretty_paths = []
    seq_net_paths = []

    for path in finished_paths:
        ppath = Path()
        seq_net = []

        for net in path :
            if stree.is_sentinel(net) :
                net = stree.get_node_from_sentinel(net)

            # case node : get the node length
            if stree.is_node(net) :
                ppath.addNodeHandle(net, stree)
                node_start_id = stree.node_id(net)
                node_handle = pg.get_handle(node_start_id)
                seq_node = pg.get_sequence(node_handle)
                seq_net.append(seq_node)

            # case trivial_chain : get the first node length
            if stree.is_trivial_chain(net) :
                ppath.addNodeHandle(net, stree)
                stn_start = stree.get_bound(net, False, True) if stree.starts_at_start(net) else stree.get_bound(net, True, True)
                node_start_id = stree.node_id(stn_start)
                net_trivial_chain = pg.get_handle(node_start_id)
                seq_trivial_chain = pg.get_sequence(net_trivial_chain)
                seq_net.append(seq_trivial_chain)

            elif stree.is_chain(net) :
                # if it's a chain, we need to write someting like ">Nl>*>Nr"
                if stree.starts_at_start(net):
                    nodl = stree.get_bound(net, False, True)
                    nodr = stree.get_bound(net, True, False)
                else:
                    nodl = stree.get_bound(net, True, True)
                    nodr = stree.get_bound(net, False, False)
                ppath.addNodeHandle(nodl, stree)
                ppath.addNode('*', '>')
                ppath.addNodeHandle(nodr, stree)
                seq_net.append("_")

        # check if path is mostly traversing nodes in reverse orientation
        if ppath.nreversed() > ppath.size() / 2 :
            ppath.flip()
            seq_net.reverse()

        pretty_paths.append(ppath.print()) 
        seq_net_paths.append(seq_net)

    type_variants, length_first_variant = calcul_pos_type_variant(seq_net_paths)
    assert len(type_variants) == len(pretty_paths)
    return pretty_paths, type_variants, length_first_variant

def loop_over_snarls_write(stree, snarls, pg, output_file, output_snarl_not_analyse, children_treshold=50, bool_return=True) :

    with open(output_file, 'w') as out_snarl, open(output_snarl_not_analyse, 'w') as out_fail:
        out_snarl.write('chr\tpos\tsnarl\tpaths\ttype\n')
        out_fail.write('snarl\treason\n')

        snarl_paths = []
        paths_number_analysis = 0
        time_threshold = 2 # 2s max per snarl analysis

        children = [0]
        def count_children(net):
            children[0] += 1
            return (True)

        # for each snarl, lists paths through the netgraph and write to output TSV
        for snarl_path_pos in snarls:
            
            snarl = snarl_path_pos[0]
            snarl_time = time.time()
            snarl_id = find_snarl_id(stree, snarl)
            not_break = True
            children = [0]

            stree.for_each_child(snarl, count_children)
            if children[0] > children_treshold :
                out_fail.write('{}\t{}\n'.format(snarl_id, "too_many_children"))
                continue

            # we'll traverse the netgraph starting at the left boundary
            # init unfinished paths to the first boundary node
            paths = [[stree.get_bound(snarl, False, True)]]
            finished_paths = []
            while len(paths) > 0 :
                path = paths.pop()
        
                if time.time() - snarl_time > time_threshold :
                    out_fail.write('{}\t{}\n'.format(snarl_id, "time_calculation_out"))
                    not_break = False
                    break

                follow_edges(stree, finished_paths, path, paths, pg)

            if not_break :

                # prepare path list to output and write each path directly to the file
                pretty_paths, type_variants, padding = fill_pretty_paths(stree, pg, finished_paths)
                out_snarl.write('{}\t{}\t{}\t{}\t{}\n'.format(snarl_path_pos[1], snarl_path_pos[2]+padding, snarl_id, ','.join(pretty_paths), ','.join(type_variants)))

                if bool_return :
                    snarl_paths.append((snarl_path_pos[1], snarl_path_pos[2]+padding, snarl_id, pretty_paths, ','.join(type_variants)))
                
                paths_number_analysis += len(pretty_paths)

    return snarl_paths, paths_number_analysis

if __name__ == "__main__" :

    parser = argparse.ArgumentParser('List path through the netgraph of each snarl in a pangenome')
    parser.add_argument('-p', type=utils.check_file, help='The input pangenome .pg file', required=True)
    parser.add_argument('-d', type=utils.check_file, help='The input distance index .dist file', required=True)
    parser.add_argument('-c', "--chr", type=utils.check_file, help='The input reference chr file', required=False)
    parser.add_argument("-t", type=check_threshold, help='Children threshold', required=False)
    parser.add_argument('-o', help='output dir', type=str, required=False)
    args = parser.parse_args()

    reference = utils.parse_chr_reference(args.chr) if args.chr else "ref"
    output_dir = args.o or "output"
    os.makedirs(output_dir, exist_ok=True)
    output = os.path.join(output_dir, "list_snarl_paths.tsv")
    output_snarl_not_analyse = os.path.join(output_dir, "snarl_not_analyse.tsv")
    stree, pg, root, pp_overlay = parse_graph_tree(args.p, args.d)

    snarls = save_snarls(stree, root, pg, reference, pp_overlay)
    print(f"Total of snarls found : {len(snarls)}")
    print("Saving snarl path decomposition...")

    threshold = args.t if args.t else 10
    _, paths_number_analysis = loop_over_snarls_write(stree, snarls, pg, output, output_snarl_not_analyse, threshold, False)
    print(f"Total of paths analyse : {paths_number_analysis}")

    # python3 stoat/list_snarl_paths.py -p /home/mbagarre/Bureau/droso_data/fly/fly.pg -d /home/mbagarre/Bureau/droso_data/fly/fly.dist -o output/test
    # vg find -x ../snarl_data/fly.gbz -r 5176878:5176884 -c 10 | vg view -dp - | dot -Tsvg -o ../snarl_data/subgraph.svg

    # binary 
    # python3 stoat/list_snarl_paths.py -p tests/simulation/binary/pg.full.pg -d tests/simulation/binary/pg.dist -c tests/simulation/binary/pg.chromosome -o output/test
    
    # quantitative
    # python3 stoat/list_snarl_paths.py -p tests/simulation/quantitative/pg.full.pg -d tests/simulation/quantitative/pg.dist -c tests/simulation/quantitative/pg.chromosome -o output/test

File: tests/src/list_snarl_paths.cpp
#include "list_snarl_paths.hpp"

using namespace std;
using namespace bdsg;
using bdsg::HandleGraph;
using handlegraph::handle_t;
using handlegraph::net_handle_t;

// Add a node with known orientation
void Path::addNode(const std::string& node, char orient) {
    nodes.push_back(node);
    orients.push_back(orient);
}

// Add a node handle and extract information using the string representation
void Path::addNodeHandle(const net_handle_t& node_h, const SnarlDistanceIndex& stree) {
    std::string node_s = stree.net_handle_as_string(node_h);

    // Handle trivial chain modifications
    if (stree.is_trivial_chain(node_h)) {
        size_t pos;
        while ((pos = node_s.find(" pretending to be a chain")) != std::string::npos) {
            node_s.replace(pos, 23, "");
        }
        while ((pos = node_s.find(" in a simple snarl")) != std::string::npos) {
            node_s.replace(pos, 19, "");
        }
    }

    // Parse node info
    size_t pos = node_s.find("node ");
    if (pos != std::string::npos) {
        node_s.erase(pos, 5);
    }

    char node_o = '>';
    if (node_s.find("rev") != std::string::npos) {
        node_o = '<';
    }

    auto removeSubstrings = [](std::string& str, const std::vector<std::string>& substrings) {
        for (const auto& sub : substrings) {
            size_t pos;
            while ((pos = str.find(sub)) != std::string::npos) {
                str.erase(pos, sub.length());
            }
        }
    };

    removeSubstrings(node_s, {"rev", "fd"});

    // Add node to path
    nodes.push_back(node_s);
    orients.push_back(node_o);
}

// Get the string representation of the path
std::string Path::print() const {
    std::string out_path;
    for (size_t i = 0; i < nodes.size(); ++i) {
        out_path += orients[i] + nodes[i];
    }
    return out_path;
}

// Flip the path orientation
void Path::flip() {
    std::reverse(nodes.begin(), nodes.end());
    std::reverse(orients.begin(), orients.end());
    for (size_t i = 0; i < orients.size(); ++i) {
        if (nodes[i] == "*") {
            continue;
        }
        orients[i] = (orients[i] == '>') ? '<' : '>';
    }
}

// Get the size of the path
size_t Path::size() const {
    return nodes.size();
}

// Count the number of reversed nodes
size_t Path::nreversed() const {
    return std::count(orients.begin(), orients.end(), '<');
}

// Function to calculate the type of variant
vector<string> calcul_type_variant(const vector<vector<string>>& list_list_length_paths) {
    vector<string> list_type_variant;

    for (const auto& path_lengths : list_list_length_paths) {
        if (path_lengths.size() > 3 || path_lengths[1] == "-1") { // Case snarl in snarl / Indel
            list_type_variant.push_back("COMPLEX");
        } else if (path_lengths.size() == 3) { // Case simple path len 3
            list_type_variant.push_back((path_lengths[1] == "1") ? "SNP" : "INS");
        } else { // Deletion
            list_type_variant.push_back("DEL");
        }
    }
    return list_type_variant;
}

// Function to check threshold
void check_threshold(double proportion) {
    if (proportion <= 0) {
        throw invalid_argument("Proportion value must be >0.");
    }
}

string find_snarl_id(SnarlDistanceIndex& stree, net_handle_t& snarl) {
    // Get start and end boundary nodes for the snarl
    auto sstart = stree.get_bound(snarl, false, true);  // False for the left boundary
    auto send = stree.get_bound(snarl, true, true);     // True for the right boundary

    // Convert the sentinels into nodes
    auto start_node = stree.get_node_from_sentinel(sstart);
    auto end_node = stree.get_node_from_sentinel(send);

    // Get the node IDs from SnarlDistanceIndex
    auto start_node_id = stree.node_id(start_node);
    auto end_node_id = stree.node_id(end_node);

    // Construct the snarl ID as "end_node_id_start_node_id"
    std::stringstream snarl_id;
    snarl_id << end_node_id << "_" << start_node_id;

    return snarl_id.str();  // Return the generated snarl ID as a string
}

void follow_edges(SnarlDistanceIndex& stree, vector<vector<net_handle_t>>& finished_paths,
                  vector<net_handle_t>& path,
                  vector<vector<net_handle_t>>& paths, PackedGraph& pg) {
}

pair<vector<tuple<string, vector<string>, string, string, string>>, int> loop_over_snarls_write(
        SnarlDistanceIndex& stree, 
        vector<tuple<net_handle_t, string, size_t>> snarls,
        PackedGraph& pg, const string& output_file,
        const string& output_snarl_not_analyse,
        int children_threshold = 50, bool bool_return = true) {

    ofstream out_snarl(output_file);
    ofstream out_fail(output_snarl_not_analyse);
    
    out_snarl << "snarl\tpaths\ttype\tchr\tpos\n";
    out_fail << "snarl\treason\n";
    
    vector<tuple<string, vector<string>, string, string, string>> snarl_paths;
    int paths_number_analysis = 0;
    chrono::seconds time_threshold(2);
    
    for (const auto& snarl_path_pos : snarls) {
        net_handle_t snarl = std::get<0>(snarl_path_pos);
        auto snarl_time = chrono::steady_clock::now();
        string snarl_id = find_snarl_id(stree, snarl);
        bool not_break = true;
        int children = 0;
        
        auto count_children = [&](auto net) {
            children++;
            return true;
        };
        
        stree.for_each_child(snarl, count_children);
        if (children > children_threshold) {
            out_fail << snarl_id << "\ttoo_many_children\n";
            continue;
        }
        
        vector<vector<net_handle_t>> paths = {{stree.get_bound(snarl, false, true)}};
        vector<vector<net_handle_t>> finished_paths;

        while (!paths.empty()) {
            auto path = paths.back();
            paths.pop_back();

            if (chrono::steady_clock::now() - snarl_time > time_threshold) {
                out_fail << snarl_id << "\ttime_calculation_out\n";
                not_break = false;
                break;
            }

            follow_edges(stree, finished_paths, path, paths, pg);
        }

        if (not_break) {
            // pair<vector<string>, vector<string>>
            auto [pretty_paths, type_variants] = fill_pretty_paths(stree, pg, finished_paths);
            std::ostringstream pretty_paths_stream, type_variants_stream;

            // Convert pretty_paths (vector<string>) into a comma-separated string
            for (size_t i = 0; i < pretty_paths.size(); ++i) {
                if (i > 0) pretty_paths_stream << ",";
                pretty_paths_stream << pretty_paths[i];
            }

            // Convert type_variants to a comma-separated string
            for (size_t i = 0; i < type_variants.size(); ++i) {
                if (i > 0) type_variants_stream << ","; // Separate vectors with commas

                // Convert inner vector (type_variants[i]) to a comma-separated string
                for (size_t j = 0; j < type_variants[i].size(); ++j) {
                    if (j > 0) type_variants_stream << " "; // Space-separated words within a vector
                    type_variants_stream << type_variants[i][j];
                }
            }

            out_snarl << snarl_id << "\t" << pretty_paths_stream.str() << "\t"
                    << type_variants_stream.str() << "\t" << std::get<1>(snarl_path_pos)
                    << "\t" << std::to_string(std::get<2>(snarl_path_pos)) << "\n";

            if (bool_return) {
                snarl_paths.emplace_back(snarl_id, pretty_paths, type_variants_stream.str(),
                                        std::get<1>(snarl_path_pos), std::to_string(std::get<2>(snarl_path_pos)));
            }

            paths_number_analysis += pretty_paths.size();
        }
    }
    
    return {snarl_paths, paths_number_analysis};
}

vector<tuple<net_handle_t, string, size_t>> save_snarls(SnarlDistanceIndex& stree, const net_handle_t& root,
                                 PackedGraph& pg, unordered_set<string>& ref_paths,
                                 PackedPositionOverlay& ppo) {
    vector<tuple<net_handle_t, string, size_t>> snarls;
    unordered_map<string, vector<string>> snarls_pos;

    // Given a node handle (dist index), return a position on a reference path
    auto get_node_position = [&](auto node) -> vector<string> { // node : net_handle_t
        handle_t node_h = stree.get_handle(node, pg);
        vector<string> ret_pos(2); // pair<string, size_t>

        pg.for_each_step_on_handle(node_h, [&](auto step_handle) {
            path_handle_t path_handle = pg.get_path_handle_of_step(step_handle);
            string path_name = pg.get_path_name(path_handle);

            if (ref_paths.count(path_name)) {
                size_t position = ppo.get_position_of_step(step_handle);
                ret_pos[0] = path_name;
                ret_pos[1] = position;
                return false; // Stop iteration once a reference path is found
            }
            return true; // Continue iteration
        });
        return ret_pos;
    };

    auto get_net_start_position = [&](net_handle_t net) -> vector<string> {
        if (stree.is_node(net)) {
            return get_node_position(net);
        }

        net_handle_t bnode1 = stree.get_bound(net, true, false);
        vector<string> bnode1_p = get_node_position(bnode1);
        net_handle_t bnode2 = stree.get_bound(net, false, false);
        vector<string> bnode2_p = get_node_position(bnode2);

        // If one of the boundaries is not on a reference path, return it
        if (bnode1_p.empty()) return bnode1_p;
        if (bnode2_p.empty()) return bnode2_p;

        assert(bnode1_p[0] == bnode2_p[0]); // Ensure they are on the same reference path

        // Return the boundary with the smaller numerical position
        return (std::stoi(bnode1_p[1]) < std::stoi(bnode2_p[1])) ? bnode1_p : bnode2_p;
    };

    function<void(net_handle_t)> save_snarl_tree_node;
    save_snarl_tree_node = [&](net_handle_t net) {
        vector<string> snarl_pos = get_net_start_position(net);
        if (snarl_pos.empty()) {
            auto par_net = stree.get_parent(net);
            snarl_pos = snarls_pos[stree.net_handle_as_string(par_net)];
        }

        snarls_pos[stree.net_handle_as_string(net)] = snarl_pos;
        if (stree.is_snarl(net)) {
            snarls.push_back(std::make_tuple(net, snarl_pos[0], std::stoull(snarl_pos[1])));
        }

        if (!stree.is_node(net) && !stree.is_sentinel(net)) {
            stree.for_each_child(net, save_snarl_tree_node);
        }
    };
    
    stree.for_each_child(root, save_snarl_tree_node);
    return snarls;
}

std::tuple<bdsg::PackedGraph, bdsg::SnarlDistanceIndex, bdsg::PackedPositionOverlay, net_handle_t>
parse_graph_tree(const std::string& pg_file, const std::string& dist_file) {
    // Load graph and snarl tree
    bdsg::PackedGraph pg;
    pg.deserialize(pg_file);
    
    bdsg::SnarlDistanceIndex stree;
    stree.deserialize(dist_file);

    bdsg::PathHandleGraph* path_graph = &pg;
    bdsg::PackedPositionOverlay pp_overlay(path_graph);
    // /Users/aliasmatis/Desktop/stoat_cxx/tests/src/list_snarl_paths.cpp:302:33: error: no matching constructor for initialization of 'bdsg::PackedPositionOverlay'
    // bdsg::PackedPositionOverlay pp_overlay(pg);

    // Get the root of the snarl tree
    auto root = stree.get_root();
    
    return std::make_tuple(pg, stree, pp_overlay, root);
}

pair<vector<string>, vector<string>> fill_pretty_paths(
    SnarlDistanceIndex& stree, PackedGraph& pg, 
    vector<vector<net_handle_t>>& finished_paths) {
    
    vector<string> pretty_paths;
    vector<vector<string>> length_net_paths;

    for (const auto& path : finished_paths) {
        Path ppath;
        vector<string> length_net;

        for (auto net : path) {
            if (stree.is_sentinel(net)) {
                net = stree.get_node_from_sentinel(net);
            }

            if (stree.is_node(net)) {
                ppath.addNodeHandle(net, stree);
                length_net.push_back(to_string(stree.node_length(net)));
            }
            else if (stree.is_trivial_chain(net)) {
                ppath.addNodeHandle(net, stree);
                auto stn_start = stree.get_bound(net, false, true);
                auto node_start_id = stree.node_id(stn_start);
                auto net_trivial_chain = pg.get_handle(node_start_id);
                length_net.push_back(to_string(pg.get_length(net_trivial_chain)));
            }
            else if (stree.is_chain(net)) {
                net_handle_t nodl, nodr;
                if (stree.starts_at_start(net)) {
                    nodl = stree.get_bound(net, false, true);
                    nodr = stree.get_bound(net, true, false);
                } else {
                    nodl = stree.get_bound(net, true, true);
                    nodr = stree.get_bound(net, false, false);
                }
                ppath.addNodeHandle(nodl, stree);
                ppath.addNode("*", '>');
                ppath.addNodeHandle(nodr, stree);
                length_net.push_back("-1");
            }
        }

        if (ppath.nreversed() > ppath.size() / 2) {
            ppath.flip();
        }
        pretty_paths.push_back(ppath.print());
        length_net_paths.push_back(length_net);
    }

    vector<string> type_variants = calcul_type_variant(length_net_paths);
    return {pretty_paths, type_variants};
}
